### PR TITLE
Add output granularity support to BaseDateTimeTransformer

### PR DESCRIPTION
**Pull request by subprotocol**
Allows date-time bucketing to be arbitrarily determined at query time.

For background discussion see: https://github.com/linkedin/pinot/issues/2729

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ protected long transformMillisToEpoch(long millisSinceEpoch) {
   }
 
   protected String transformMillisToSDF(long millisSinceEpoch) {
-    return _outputDateTimeFormatter.print(millisSinceEpoch);
+    return _outputDateTimeFormatter.print(transformToOutputGranularity(millisSinceEpoch));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java]: Should not embed **transformToOutputGranularity** into **transformMillisToSDF**, but add it into **Epoch/SDFToSDFTransformer** (refer to **SDFToEpochTransformer**)

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
 /**
  * Base date time transformer to transform and bucket date time values from epoch/simple date format to epoch/simple
  * date format.
- * <p>NOTE: time size, time unit and output granularity does not apply to simple date format.
+ * <p>NOTE: time size and time unit do not apply to simple date format.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java]: Thanks for updating the javadoc

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java (hunk 1 above)
Comment (subprotocol) [reply pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java]: Ahh I see now thank you, I will get that changed.

**Comment by kishoreg** [comment]
@npawar raised a good point about handling timezones in #2729. Lets make sure we handle that.

**Comment by subprotocol** [comment]
any chance this could get a quick review?  I have some cycles coming up that if needed I can make additional improvements if required.

**Comment by npawar** [comment]
LGTM as well. 
There were a lot of test cases at one point which we lost sometime back (https://github.com/linkedin/pinot/pull/2659/files#diff-7b9f15468a245a32ab97a2f6bd1f0af9)
Jackie is adding them back. If you could hold on until those are in, and verify that your change passes them all, that would be great

**Comment by subprotocol** [comment]
Sure no problem, that sounds like a good a good plan. Thanks!

**Comment by npawar** [comment]
Tests have been added back @subprotocol

**Comment by subprotocol** [comment]
Thank you!  I will take a look at this today

**Comment by subprotocol** [comment]
@npawar ok all merged up.  I updated the tests I had written to follow the flow of the tests that were put back into DateTimeConverterTest.

**Comment by npawar** [comment]
Since your change is in millisToSDF, please also add some tests for SDF to SDF conversion, since SDFToSDFTransform is also affected by the change

**Comment by subprotocol** [comment]
It's a slight tangent from the current path, but one thing I wanted to raise- should timeunits even be allowed when using SIMPLE_DATE_FORMAT?

Currently a time unit is required on the inputFmt and outputFmt. ie-
```1:MILLISECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd HH:mm:ss.SSS tz(America/Los_Angeles)```

Should we update the DateTimeFormatSpec to not require the timeunit when using SIMPLE_DATE_FORMAT?

**Comment by subprotocol** [comment]
:thumbsup:

**Comment by subprotocol** [comment]
thanks!

**Pull request merged**